### PR TITLE
Stack can create docker image(s) from your executable(s)

### PR DESCRIPTION
**Pull request by dysinger**
There is still more to add to this but it works (outside of docker-in-docker)

### CODE REVIEW

**Comment by dysinger** [comment]
docker in docker works too as soon as stack-image-builder is updated

**Comment by cocreature** [comment]
I was just trying this out and run into a small bug (at least I consider it to be one):
You need to run `stack install` before `stack image container`, otherwise it fails with `copyFile: does not exist (No such file or directory)`.

I don't really see why it should be required to install the executables globally. If there is a good reason for that this should happen automatically or at least provide a better error message.

**Comment by cocreature** [comment]
When I specify `snoyberg/haskell-scratch` as my base image then my executables don't seem to be contained in the image. Changing this to `fpco/stack-build` and using the same path for the executables works just fine. No idea what could be causing this.

Also a feature request (I can open another issue for that if you want):

I'd like to have different containers for my executables. The rational behind this is that I put packages in one project that somehow logically fit together (e.g. a web scraper and a webservice presenting the scraped data) but imho it really would make more sense to have separate images for this. Obviously I could have one image and then use different containers which start separate executables but since they don't really share anything I feel like having separate images here is better.

File: src/Path/IO.hs [hunk 1]
```diff
@@ -224,6 +226,12 @@ copyDirectoryRecursive srcDir destDir =
                   Nothing -> return ()
                   Just relSubDir -> copyDirectoryRecursive srcSubDir (destDir </> relSubDir))
 
+
+-- | Drop the root (either @\/@ on POSIX or @C:\\@, @D:\\@, etc. on
+-- Windows).
+dropRoot :: Path Abs t -> Path Rel t
```
Comment (manny-fp) [inline src/Path/IO.hs]: This seems like it might better belong in Path itself (since it does no I/O).

**Comment by manny-fp** [comment]
LGTM, so as far as I'm concerned go ahead and merge it whenever you're ready.

**Comment by manny-fp** [comment]
> I'd like to have different containers for my executables [...]
@cocreature please file an "enhancement" issue.

**Comment by dysinger** [comment]
@cocreature please do criticize :) I have thick skin & could use helpful critique.  Also this docker stuff is tricky.  I'm not completely satisfied with this effort as of today.

File: src/Path/IO.hs (hunk 1 above)
Comment (dysinger) [inline src/Path/IO.hs]: I know.  Chris suggested that I put it there for now. (slack chat) He said it was a staging area for possible Path additions.

**Comment by mgsloan** [comment]
> I was just trying this out and run into a small bug (at least I consider it to be one):
You need to run stack install before stack image container, otherwise it fails with copyFile: does not exist (No such file or directory).

@cocreature , I ran into this as well - fixed here https://github.com/commercialhaskell/stack/commit/c028b81f478ecabef753c64d54b11952533b2021

**Comment by dysinger** [comment]
thanks @mgsloan

**Comment by fizruk** [comment]
Is there any documentation for this anywhere?

Also, I believe this has solved #401 (so I am adding this reference, because there was none and it took me some time to find this PR).

**Comment by fizruk** [comment]
Found documentation [here](https://github.com/commercialhaskell/stack/wiki/stack.yaml#image).

**Comment by nicemaker** [comment]
I got yesterday a docker image running, build with stack. But I have to call it like 
`docker run -p 5000:5000 {imagename} {app-exe}`.

What I really want is that the app-exe is started automatically, I thought `endpoint` is suppose to do that. I tried

```
endpoint:
    - app-exe
```

That doesn't work for me unfortunately. Am I missunderstanding entrypoint?

Stack 1.3.2
Docker  for Mac

**Pull request merged**